### DISCUSSION TITLE
Proposal for a new CLI argument --continue-on-error or --stop-on-error

### DISCUSSION THREAD

**Discussion by minghungcho** [littleK0i/SnowDDL#11] [Ideas]
Currently SnowDDL continues execution even when encountering an error. It results partially executed DDL scripts.

ex.
```
2022-06-30 22:51:19.506 - WARNING - Resolved TABLE [SAMPLE_DB.PUBLIC.EMPLOYEE]: ERROR
(
    message   =>  SQL compilation error: error line 3 at position 41
....
```
Unfortunately, this doesn't work well in a CI/CD pipeline when the process should have been stopped so that the issue can be addressed before propagated to the next environment.

Similar to [Azure Task option](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/tasks?view=azure-devops&tabs=yaml#task-control-options), a new CLI argument --continue-on-error or --stop-on-error will prevent SnowDDL from executing any DDL script when any issues found during Resolvers (ex. insufficient permission, SQL compilation error, ..).

This will allow CI/CD pipeline to halt properly, and given batch script a chance to exit gracefully. Maybe even returned a system exit code?

Personally, I like the an option **--continue-on-error** to explicitly instruct SnowDDL to continue its old behavior. However, for others who are accustomed to its current behavior, maybe an argument like **--stop-on-error** would be less disruptive.

BTW, a colleague of mine found the above "WARNING .. ERROR" message confusing, which I also agree. I guess because it shows as a WARNING, SnowDDL decides to continue its execution even the detailed message says it is an ERROR?

**Comment by littleK0i** [comment]
There are few reasons why SnowDDL currently resolves all blueprints and does not stop on errors:

1. Objects of the same type are resolved not sequentially, but in parallel, in separate threads. Also, the order of execution is unpredictable due to variation of execution time in Snowflake. This approach is necessary for performance.
2. Snowflake does not support transactions for DDL queries. Everything is committed immediately, and DDL queries cannot be reversed easily.

It means you will always have a chance to get multiple errors due to parallelism, and you may also experience "broken" state in Snowflake account. Stopping the execution will not prevent these fundamental problems.

What I can do is add `--stop-on-error` option, which can stop execution between object types in case of errors. For example, if some `TABLES` produced SQL error, SnowDDL will stop, and `VIEWS` will not be resolved.

Also, I can add a special non-zero exitcode if any execution errors were detected. It should help CI/CD pipelines to detect errors without parsing output.

---

About `WARNING ... ERROR`.

`ERROR` is a part of `ResolveResult` enum: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/resolver/abc_resolver.py#L15-L24

Resolution always ends with one of `ResolveResult`. `ERROR` result means SQL ERROR occurred while SnowDDL tried to resolve an object.

But WARNING is a different thing. It is a log level for log message. It means SnowDDL tries to resolve an object, caught the error, displayed log line and continued execution.

**Comment by littleK0i** [comment]
On a second thought, even sequential execution would not help.

Sometimes it is required to execute multiple DDL queries during resolution of one object. For example, `ROLE` can be resolved as:

```
CREATE ROLE <name>;
GRANT <something> TO ROLE <name>;
GRANT <something_else> TO ROLE <name>;
```

You may get an exception somewhere in the middle, and role will get stuck between "previous" and "desired" state anyway.

And if "invalid" states are unavoidable, regardless of object management tool, it makes sense to try and resolve as much as possible, and see all errors in one go. So this was the thought process behind the original implementation.

**Reply by minghungcho** [reply]
I see where the challenge is.
Since SnowDDL is stateless, I can always do a dry-run against a clone before applying it to the target database. Actually, this might be an interesting feature to add to SnowDDL later!

I think If SnowDDL would return an error code, I can halt Azure pipeline properly.

**Reply by littleK0i** [reply]
@minghungcho , since version `0.7.3`, SnowDDL returns exit code `8` if any errors occurred inside resolvers.

Since it is non-zero exit code, it should stop CI/CD pipelines properly.

Hope it helps!
